File: java/GrepMtTuned.java
// jgrep_std_mt_tuned - concurrent Java with the memory pillar applied:
// each worker reuses ONE growable read buffer (and one lowercase buffer) across
// files; reads a 64 KB prefix first, does the NUL binary-check on that prefix,
// and only reads the rest of the file if it passed. Mirrors go/mt/main.go.
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicInteger;

public class GrepMtTuned {
    static final int PEEK = 65536;

    static byte[] pat;
    static byte[] lpat;
    static boolean patSet;
    static boolean ci;
    static boolean recursive;
    static boolean multi;

    static final java.util.concurrent.atomic.AtomicBoolean anyMatch =
        new java.util.concurrent.atomic.AtomicBoolean(false);
    static OutputStream out;
    static final Object outLock = new Object();
    static final ArrayList<String> files = new ArrayList<>();
    static final AtomicInteger idx = new AtomicInteger(0);

    static void asciiLower(byte[] dst, byte[] src, int len) {
        for (int i = 0; i < len; i++) {
            byte b = src[i];
            if (b >= 'A' && b <= 'Z') dst[i] = (byte) (b + 32);
            else dst[i] = b;
        }
    }

    static int indexOf(byte[] hay, int len, int from, byte[] needle) {
        int n = needle.length;
        if (n == 0) return from;
        int end = len - n;
        for (int i = from; i <= end; i++) {
            int j = 0;
            while (j < n && hay[i + j] == needle[j]) j++;
            if (j == n) return i;
        }
        return -1;
    }

    static int lastIndexOfNL(byte[] data, int m) {
        for (int i = m - 1; i >= 0; i--) if (data[i] == '\n') return i;
        return -1;
    }

    static int indexOfNL(byte[] data, int from, int len) {
        for (int i = from; i < len; i++) if (data[i] == '\n') return i;
        return -1;
    }

    // Per-worker reusable scratch.
    static final class Scratch {
        byte[] rbuf = new byte[PEEK];   // growable read buffer, reused across files
        byte[] lowbuf = new byte[0];    // growable lowercase scratch
        ByteArrayOutputStream w = new ByteArrayOutputStream(1 << 16);
    }

    // read at most cap-off bytes into buf[off..]; returns total length in buf.
    static int readFully(InputStream in, byte[] buf, int off, int want) throws IOException {
        int n = off;
        int endTarget = off + want;
        while (n < endTarget) {
            int r = in.read(buf, n, endTarget - n);
            if (r < 0) break;
            n += r;
        }
        return n;
    }

    static boolean searchFile(String path, Scratch s) throws IOException {
        Path pp = Paths.get(path);
        long sizeL;
        try {
            BasicFileAttributes a = Files.readAttributes(pp, BasicFileAttributes.class);
            sizeL = a.size();
        } catch (IOException e) {
            return false;
        }
        if (sizeL <= 0) return false;
        int size = (int) Math.min(sizeL, Integer.MAX_VALUE);
        int peek = Math.min(size, PEEK);

        try (InputStream in = Files.newInputStream(pp)) {
            if (s.rbuf.length < peek) s.rbuf = new byte[peek];
            int got = readFully(in, s.rbuf, 0, peek);
            if (got < peek) peek = got; // short read; work with what we have
            for (int i = 0; i < peek; i++) if (s.rbuf[i] == 0) return false; // binary

            int len = peek;
            if (size > peek) {
                if (s.rbuf.length < size) {
                    byte[] nb = new byte[size];
                    System.arraycopy(s.rbuf, 0, nb, 0, peek);
                    s.rbuf = nb;
                }
                len = readFully(in, s.rbuf, peek, size - peek);
            }
            byte[] data = s.rbuf;

            byte[] hay = data;
            byte[] needle = pat;
            if (ci) {
                if (s.lowbuf.length < len) s.lowbuf = new byte[len];
                asciiLower(s.lowbuf, data, len);
                hay = s.lowbuf;
                needle = lpat;
            }
            byte[] pathBytes = path.getBytes(java.nio.charset.StandardCharsets.ISO_8859_1);
            boolean found = false;
            int pos = 0;
            // pos < len: empty needle matches each line once (grep -F "" semantics).
            while (pos < len) {
                int m = indexOf(hay, len, pos, needle);
                if (m < 0) break;
                int ls = lastIndexOfNL(data, m) + 1;
                int j = indexOfNL(data, m, len);
                int le = (j >= 0) ? j : len;
                found = true;
                if (multi) {
                    s.w.write(pathBytes, 0, pathBytes.length);
                    s.w.write(':');
                }
                s.w.write(data, ls, le - ls);
                s.w.write('\n');
                pos = le + 1;
            }
            return found;
        } catch (IOException e) {
            return false;
        }
    }

    static void collect(Path dir) {
        try {
            Files.walkFileTree(dir, new java.nio.file.SimpleFileVisitor<Path>() {
                @Override
                public java.nio.file.FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
                    if (attrs.isRegularFile()) files.add(file.toString());
                    return java.nio.file.FileVisitResult.CONTINUE;
                }
                @Override
                public java.nio.file.FileVisitResult visitFileFailed(Path file, IOException exc) {
                    return java.nio.file.FileVisitResult.CONTINUE;
                }
            });
        } catch (IOException e) { /* ignore */ }
    }

    static void usage() {
        System.err.print("usage: jgrep [-r] [-i] PATTERN PATH...\n");
        System.exit(2);
    }

    public static void main(String[] args) throws Exception {
        ArrayList<String> paths = new ArrayList<>();
        boolean noMore = false;
        for (String a : args) {
            if (!noMore && a.length() >= 2 && a.charAt(0) == '-') {
                if (a.equals("--")) { noMore = true; continue; }
                for (int k = 1; k < a.length(); k++) {
                    char c = a.charAt(k);
                    if (c == 'i') ci = true;
                    else if (c == 'r') recursive = true;
                    else usage();
                }
            } else if (!patSet) {
                pat = a.getBytes(java.nio.charset.StandardCharsets.ISO_8859_1);
                patSet = true;
            } else {
                paths.add(a);
            }
        }
        if (!patSet || paths.isEmpty()) usage();
        lpat = new byte[pat.length];
        asciiLower(lpat, pat, pat.length);
        multi = recursive || paths.size() > 1;

        for (String p : paths) {
            Path pp = Paths.get(p);
            BasicFileAttributes attrs;
            try {
                attrs = Files.readAttributes(pp, BasicFileAttributes.class);
            } catch (IOException e) {
                continue;
            }
            if (attrs.isDirectory()) {
                if (recursive) collect(pp);
            } else {
                files.add(p);
            }
        }

        out = new BufferedOutputStream(System.out, 1 << 16);
        int nthreads = Runtime.getRuntime().availableProcessors();
        Thread[] threads = new Thread[nthreads];
        for (int t = 0; t < nthreads; t++) {
            threads[t] = new Thread(() -> {
                Scratch s = new Scratch();
                try {
                    while (true) {
                        int i = idx.getAndIncrement();
                        if (i >= files.size()) break;
                        s.w.reset();
                        boolean f = searchFile(files.get(i), s);
                        if (f) anyMatch.set(true);
                        if (s.w.size() > 0) {
                            synchronized (outLock) {
                                s.w.writeTo(out);
                            }
                        }
                    }
                } catch (IOException e) { /* ignore */ }
            });
            threads[t].start();
        }
        for (Thread th : threads) th.join();
        out.flush();
        System.exit(anyMatch.get() ? 0 : 1);
    }
}

File: kotlin/grep_mt_tuned.kt
// ktgrep_std_mt_tuned - concurrent Kotlin, memory-tuned. Each worker thread
// reuses ONE growable read buffer (and one lowercase buffer) across files.
// Reads only a 64 KB prefix first, does the NUL binary-check on that prefix,
// and reads the rest of the file only if it passed. Workers pull file indices
// off an atomic counter; each file's output block is written under a lock.
import java.io.BufferedOutputStream
import java.io.File
import java.io.RandomAccessFile
import java.util.concurrent.atomic.AtomicInteger
import kotlin.concurrent.thread
import kotlin.system.exitProcess

private var pat = ByteArray(0)
private var lpat = ByteArray(0)
private var ci = false
private var recursive = false
private var multi = false

private fun asciiLower(dst: ByteArray, src: ByteArray, n: Int) {
    for (i in 0 until n) {
        val b = src[i]
        dst[i] = if (b >= 'A'.code.toByte() && b <= 'Z'.code.toByte()) (b + 32).toByte() else b
    }
}

private fun indexOf(hay: ByteArray, hayLen: Int, needle: ByteArray, from: Int): Int {
    val n = needle.size
    if (n == 0) return from
    if (from + n > hayLen) return -1
    val first = needle[0]
    var i = from
    val last = hayLen - n
    while (i <= last) {
        if (hay[i] == first) {
            var j = 1
            while (j < n && hay[i + j] == needle[j]) j++
            if (j == n) return i
        }
        i++
    }
    return -1
}

private fun lastNl(data: ByteArray, end: Int): Int {
    var i = end - 1
    while (i >= 0) {
        if (data[i] == '\n'.code.toByte()) return i + 1
        i--
    }
    return 0
}

// Per-worker reusable scratch buffers.
private class Scratch {
    var rbuf = ByteArray(65536)   // growable read buffer, reused across files
    var lowbuf = ByteArray(0)     // growable lowercase buffer, reused across files
}

private fun growKeep(buf: ByteArray, need: Int, keep: Int): ByteArray {
    if (buf.size >= need) return buf
    val nb = ByteArray(need)
    System.arraycopy(buf, 0, nb, 0, keep)
    return nb
}

// Returns true if any match found; appends output bytes into `w`.
private fun searchFile(path: String, w: StringBuilder, s: Scratch): Boolean {
    val raf = try {
        RandomAccessFile(path, "r")
    } catch (e: Exception) {
        return false
    }
    try {
        val sizeL = raf.length()
        if (sizeL <= 0L) return false
        val size = if (sizeL > Int.MAX_VALUE) Int.MAX_VALUE else sizeL.toInt()
        val peek = if (size > 65536) 65536 else size

        if (s.rbuf.size < peek) s.rbuf = growKeep(s.rbuf, peek, 0)
        raf.readFully(s.rbuf, 0, peek)

        for (i in 0 until peek) {
            if (s.rbuf[i] == 0.toByte()) return false // binary: rest unread
        }

        if (size > peek) {
            s.rbuf = growKeep(s.rbuf, size, peek) // grow but preserve the prefix
            raf.readFully(s.rbuf, peek, size - peek)
        }
        val data = s.rbuf
        val len = size

        val hay: ByteArray
        val needle: ByteArray
        if (ci) {
            if (s.lowbuf.size < len) s.lowbuf = ByteArray(len)
            asciiLower(s.lowbuf, data, len)
            hay = s.lowbuf
            needle = lpat
        } else {
            hay = data
            needle = pat
        }

        var found = false
        // `pos < len` (not `<=`): with `<=`, an empty needle emits a spurious
        // blank match past the trailing newline. grep -F: empty pattern == one
        // match per line. A non-empty needle can never match at pos==len.
        var pos = 0
        while (pos < len) {
            val m = indexOf(hay, len, needle, pos)
            if (m < 0) break
            val ls = lastNl(data, m)
            var le = len
            var j = m
            while (j < len) {
                if (data[j] == '\n'.code.toByte()) { le = j; break }
                j++
            }
            found = true
            if (multi) {
                w.append(path)
                w.append(':')
            }
            for (k in ls until le) w.append((data[k].toInt() and 0xFF).toChar())
            w.append('\n')
            pos = le + 1
        }
        return found
    } catch (e: Exception) {
        return false
    } finally {
        try { raf.close() } catch (e: Exception) {}
    }
}

private fun usage(): Nothing {
    System.err.print("usage: ktgrep [-r] [-i] PATTERN PATH...\n")
    exitProcess(2)
}

private fun collect(p: String, files: MutableList<String>) {
    val f = File(p)
    if (!f.exists()) return
    if (f.isDirectory) {
        if (recursive) {
            val stack = ArrayDeque<File>()
            stack.addLast(f)
            while (stack.isNotEmpty()) {
                val d = stack.removeLast()
                val kids = d.listFiles() ?: continue
                for (k in kids) {
                    val isSymlink = try {
                        k.canonicalFile != k.absoluteFile
                    } catch (e: Exception) { false }
                    if (isSymlink) continue
                    if (k.isDirectory) stack.addLast(k)
                    else if (k.isFile) files.add(k.path)
                }
            }
        }
    } else if (f.isFile) {
        files.add(p)
    }
}

fun main(args: Array<String>) {
    val paths = ArrayList<String>()
    var patSet = false
    var noMore = false
    for (a in args) {
        if (!noMore && a.length >= 2 && a[0] == '-') {
            if (a == "--") { noMore = true; continue }
            for (idx in 1 until a.length) {
                when (a[idx]) {
                    'i' -> ci = true
                    'r' -> recursive = true
                    else -> usage()
                }
            }
        } else if (!patSet) {
            pat = a.toByteArray(Charsets.ISO_8859_1)
            patSet = true
        } else {
            paths.add(a)
        }
    }
    if (!patSet || paths.isEmpty()) usage()
    lpat = ByteArray(pat.size)
    asciiLower(lpat, pat, pat.size)
    multi = recursive || paths.size > 1

    val files = ArrayList<String>()
    for (p in paths) collect(p, files)

    val out = BufferedOutputStream(System.out, 1 shl 16)
    val lock = Object()
    val idx = AtomicInteger(0)
    val nthreads = Runtime.getRuntime().availableProcessors()
    val workers = ArrayList<Thread>(nthreads)
    val matchFlag = booleanArrayOf(false)
    for (t in 0 until nthreads) {
        workers.add(thread(start = true) {
            val w = StringBuilder()
            val s = Scratch()
            while (true) {
                val i = idx.getAndIncrement()
                if (i >= files.size) break
                w.setLength(0)
                if (searchFile(files[i], w, s)) matchFlag[0] = true
                if (w.isNotEmpty()) {
                    val bytes = w.toString().toByteArray(Charsets.ISO_8859_1)
                    synchronized(lock) { out.write(bytes) }
                }
            }
        })
    }
    for (wkr in workers) wkr.join()
    out.flush()
    exitProcess(if (matchFlag[0]) 0 else 1)
}
